src/main/java/net/minestom/server/event/Event.java

```
@@ -1,11 +1,94 @@
+        private final Consumer<T> combined;
+
+        private Listener(Class<T> eventType, @NotNull Consumer<T> combined) {
+            this.eventType = eventType;
+            this.combined = combined;
+        }
+
+        public void attachTo(@NotNull EventHandler handler) {
+            final boolean success = this.attach.add(handler);
+            if (success) {
+                handler.addEventCallback(eventType, combined::accept);
+            }
+        }
+
+        public void detachFrom(@NotNull EventHandler handler) {
+            final boolean success = this.attach.remove(handler);
+            if (success) {
+                handler.removeEventCallback(eventType, combined::accept);
+            }
+        }
+
+        public static class Builder<T extends Event> {
+
+            private final Class<T> eventType;
+
+            private List<Function<T, Boolean>> filters = new ArrayList<>();
+            private Consumer<T> handler;
+
+            private Builder(Class<T> eventType) {
+                this.eventType = eventType;
+            }
+
+            public Builder<T> filter(Function<T, Boolean> filter) {
+                this.filters.add(filter);
+                return this;
+            }
+
+            public Builder<T> handler(Consumer<T> handler) {
+                this.handler = handler;
+                return this;
+            }
+
+            public Listener<T> build() {
+                return new Listener<>(eventType, event -> {
+                    // Filtering
+                    if (!filters.isEmpty()) {
+                        if (filters.stream().anyMatch(tBooleanFunction -> !tBooleanFunction.apply(event))) {
+                            // Cancelled
+                            return;
+                        }
+                    }
+
+                    // Handler
+                    if (handler != null) {
+                        handler.accept(event);
```

Do we want to allow people to create listeners with no actual handler?

---

I assume that we could throw an exception, but just doing nothing doesn't seem like a terrible solution either. Perhaps a warning

---

I think throwing an exception might be the best idea - I can think of zero reason to create a listener that doesn't do anything so it's most likely a developer mistake.